cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerHttpHandlerInternal.java

```diff
@@ -0,0 +1,182 @@
+import io.netty.handler.codec.http.EmptyHttpHeaders;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Internal {@link HttpHandler} for Task worker.
+ */
+@Singleton
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/system")
+public class SystemWorkerHttpHandlerInternal extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(SystemWorkerHttpHandlerInternal.class);
+  private static final Gson GSON = new GsonBuilder().registerTypeAdapter(BasicThrowable.class,
+      new BasicThrowableCodec()).create();
+  private static final String SUCCESS = "success";
+  private static final String FAILURE = "failure";
+  private final int requestLimit;
+
+  private final RunnableTaskLauncher runnableTaskLauncher;
+
+  /**
+   * Holds the total number of requests that have been executed by this handler that should count toward max allowed.
+   */
+  private final AtomicInteger requestProcessedCount = new AtomicInteger(0);
+
+  private final MetricsCollectionService metricsCollectionService;
+
+  public SystemWorkerHttpHandlerInternal(CConfiguration cConf, MetricsCollectionService metricsCollectionService) {
+    this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
+    this.metricsCollectionService = metricsCollectionService;
+    this.requestLimit = cConf.getInt(Constants.SystemWorker.REQUEST_LIMIT);
+  }
+
+  private void emitMetrics(TaskDetails taskDetails) {
+    long time = System.currentTimeMillis() - taskDetails.getStartTime();
+    Map<String, String> metricTags = new HashMap<>();
+    metricTags.put(Constants.Metrics.Tag.CLASS, taskDetails.getClassName());
+    metricTags.put(Constants.Metrics.Tag.STATUS, taskDetails.isSuccess() ? SUCCESS : FAILURE);
+    metricsCollectionService.getContext(metricTags).increment(Constants.Metrics.SystemWorker.REQUEST_COUNT, 1L);
+    metricsCollectionService.getContext(metricTags).gauge(Constants.Metrics.SystemWorker.REQUEST_LATENCY_MS, time);
+  }
+
+  @POST
+  @Path("/run")
+  public void run(FullHttpRequest request, HttpResponder responder) {
+    if (requestProcessedCount.incrementAndGet() > requestLimit) {
+      responder.sendStatus(HttpResponseStatus.TOO_MANY_REQUESTS);
```

Comment:
We shouldn't need such logic for system pods

Reply:
We are thinking of limiting the number of requests in system pods as a defense against over utilization of resources. Feel free to point out if there's a different way of approaching this.